mlflow/tracing/export/inference_table.py

```diff
@@ -0,0 +1,94 @@
+import json
+import logging
+from datetime import datetime
+from typing import Any, Dict, Optional, Sequence
+
+from cachetools import TTLCache
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExporter
+
+from mlflow.entities.span import LiveSpan
+from mlflow.environment_variables import (
+    MLFLOW_TRACE_BUFFER_MAX_SIZE,
+    MLFLOW_TRACE_BUFFER_TTL_SECONDS,
+)
+from mlflow.tracing.constant import TRACE_SCHEMA_VERSION_KEY
+from mlflow.tracing.trace_manager import InMemoryTraceManager
+
+_logger = logging.getLogger(__name__)
+
+
+def get_completed_trace(request_id: str) -> Optional[Dict[str, Any]]:
+    """
+    Get the completed trace data from the buffer. This method is used in
+    the Databricks model serving so please be careful when modifying it.
+    """
+    return _TRACE_BUFFER.get(request_id)
+
+
+# For Inference Table, we use special TTLCache to store the finished traces
+# so that they can be retrieved by the Databricks model serving. This needs
+# to be a different from the queue-based buffer used by MLflow fluent API as
+# the access is not index-based but key (request ID) based.
+_TRACE_BUFFER = TTLCache(
+    maxsize=MLFLOW_TRACE_BUFFER_TTL_SECONDS.get(),
+    ttl=MLFLOW_TRACE_BUFFER_MAX_SIZE.get(),
+)
+
+
+class InferenceTableSpanExporter(SpanExporter):
+    """
+    An exporter implementation that logs the traces to Inference Table.
+
+    Currently the Inference Table does not use collector to receive the traces,
+    but rather actively fetches the trace during the prediction process. In the
+    future, we may consider using collector-based approach and this exporter should
+    send the traces instead of storing them in the buffer.
+    """
+
+    def __init__(self):
+        self._trace_manager = InMemoryTraceManager.get_instance()
+
+    def export(self, root_spans: Sequence[ReadableSpan]):
+        """
+        Export the spans to MLflow backend.
+
+        Args:
+            spans: A sequence of OpenTelemetry ReadableSpan objects to be exported.
+                Only root spans for each trace are passed to this method.
+        """
+        for span in root_spans:
+            if span._parent is not None:
+                _logger.debug("Received a non-root span. Skipping export.")
+                continue
+
+            trace = self._trace_manager.pop_trace(span.context.trace_id)
+            if trace is None:
+                _logger.debug(f"Trace for span {span} not found. Skipping export.")
+                continue
+
+            # Add the trace to the in-memory buffer so it can be retrieved by upstream
+            _TRACE_BUFFER[trace.info.request_id] = {
+                TRACE_SCHEMA_VERSION_KEY: 2,
```

Comment:
If this isn't the only place we'll be defining the schema version, it might be worthwhile to put this into a trace versioning module for simplicity's sake

Reply:
True, eventually this should be added as a tag for any traces not only inference table. It may be just a constant and setter method in trace manager. Could you elaborate what you imagine for `trace versioning module`?

Reply:
Just something similar to what we do for MLflow version wherein a version constant is defined in a single place. Burying the version within module implementations simply runs the risk of missing one when updating :)

Reply:
Makes sense, lemme use a constant as its simplest form for now, as we don't have any method like `is_release_version`.